temporal-sdk/src/main/java/io/temporal/common/converter/CodecDataConverter.java

```diff
@@ -100,21 +106,162 @@ public <T> T fromPayloads(
       int index, Optional<Payloads> content, Class<T> valueType, Type valueGenericType)
       throws DataConverterException {
     if (content.isPresent()) {
-      List<Payload> decodedPayloads = codec.decode(content.get().getPayloadsList());
-      content = Optional.of(Payloads.newBuilder().addAllPayloads(decodedPayloads).build());
+      content = Optional.of(decode(content.get()));
     }
     return dataConverter.fromPayloads(index, content, valueType, valueGenericType);
   }
 
+  @Override
+  public Failure exceptionToFailure(Throwable e) {
+    return this.encodeFailure(dataConverter.exceptionToFailure(e));
```

Comment:
This should be configurable. I totally see users who may want to encode normal payloads and leave Failures unencoded. Pretty much have what they have currently.

Reply:
+1 this is how go failure converters work

Reply:
Oh I see it is configurable it is just in the `encodeFailure` function